src/pdn/src/PdnGen.cc

```diff
@@ -78,131 +75,96 @@ void PdnGen::init(dbDatabase* db, Logger* logger)
+void PdnGen::reportGlobalConnect(dbBlock* block)
 {
-  dbBox* bbox = nullptr;
-  globalConnectRegion(block, bbox, instPattern, pinPattern, net);
+  auto connects = block->getGlobalConnects();
+
+  logger_->report("Global connection rules: {}", connects.size());
+  for (auto* connect : connects) {
+    auto* region = connect->getRegion();
+    auto* net = connect->getNet();
+
+    std::string region_msg = "";
+    if (region != nullptr) {
+      region_msg = fmt::format(" in \"{}\"", region->getName());
+    }
+    logger_->report("  Instances{}: \"{}\" with pins \"{}\" connect to \"{}\"",
+                    region_msg,
+                    connect->getInstPattern(),
+                    connect->getPinPattern(),
+                    net->getName());
+  }
 }
 
 void PdnGen::globalConnect(dbBlock* block)
 {
-  if (global_connect_ == nullptr) {
+  auto connects = block->getGlobalConnects();
+
+  if (connects.empty()) {
     logger_->warn(PDN, 50, "Global connections not set up.");
     return;
   }
 
-  // Do non-regions first
-  if (global_connect_->find(nullptr) != global_connect_->end()) {
-    globalConnectRegion(block, nullptr, global_connect_->at(nullptr));
+  // only search for instances once
+  std::map<std::string, std::set<dbInst*>> inst_map;
+  for (auto* connect : connects) {
+    const std::string inst_pattern = connect->getInstPattern();
+    if (inst_map.find(inst_pattern) != inst_map.end()) {
+      continue;
+    }
+    inst_map[inst_pattern] = connect->getInsts();
   }
 
-  // Do regions
-  for (auto& [region, net_regex_pairs] : *global_connect_) {
-    if (region == nullptr)
-      continue;
-    globalConnectRegion(block, region, net_regex_pairs);
+  for (auto* connect : connects) {
+    globalConnect(connect, inst_map[connect->getInstPattern()]);
   }
 }
 
-void PdnGen::globalConnectRegion(dbBlock* block,
-                                 dbBox* region,
-                                 std::shared_ptr<netRegexPairs> global_connect)
+void PdnGen::globalConnect(dbGlobalConnect* global_connect)
 {
-  for (auto& [net, regex_pairs] : *global_connect) {
-    for (auto& [inst_regex, pin_regex] : *regex_pairs) {
-      globalConnectRegion(block, region, inst_regex, pin_regex, net);
-    }
+  if (global_connect == nullptr) {
+    return;
   }
+
+  globalConnect(global_connect, global_connect->getInsts());
 }
 
-void PdnGen::globalConnectRegion(dbBlock* block,
-                                 dbBox* region,
-                                 std::shared_ptr<regex>& instPattern,
-                                 std::shared_ptr<regex>& pinPattern,
-                                 dbNet* net)
+void PdnGen::globalConnect(dbGlobalConnect* global_connect, const std::set<dbInst*> insts)
 {
-  if (net == nullptr) {
-    logger_->warn(PDN, 60, "Unable to add invalid net.");
+  if (insts.empty()) {
     return;
   }
 
+  auto* net = global_connect->getNet();
   if (net->isDoNotTouch()) {
     return;
   }
```

Comment:
It's possible to change the donttouch later (either set it or clear it), the `if` will always be needed. I think generating a warning might be better since the intent might be to have the rule for earlier or later in the run.